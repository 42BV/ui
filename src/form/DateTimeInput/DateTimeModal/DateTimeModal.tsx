import { ReactNode, useState } from 'react';
import Datetime from 'react-datetime';
import moment, { Moment, MomentInput } from 'moment';
import { constant, get } from 'lodash';

import { DateTimeInputIsDateAllowed } from '../DateTimeInput';
import { DateFormat, TimeFormat } from '../types';
import { t } from '../../../utilities/translation/translation';
import { OpenCloseModal } from '../../../core/OpenCloseModal/OpenCloseModal';

export type Text = {
  cancel?: string;
  select?: string;
};

type Props = {
  /**
   * Callback for when the modal should close.
   */
  onClose: () => void;

  /**
   * The label of the form element.
   */
  label?: ReactNode;

  /**
   * Callback for when the form element changes.
   */
  onSave: (value: Moment | string) => void;

  /**
   * The value that the form element currently has.
   */
  defaultValue?: MomentInput;

  /**
   * The format for the date, follows Moment.js format.
   *
   * At least a DateFormat or TimeFormat should be defined, otherwise
   * an error occurs.
   *
   * @see https://momentjs.com/docs/#/displaying/format/
   */
  dateFormat: DateFormat;

  /**
   * The format for the time, follows Moment.js format.
   *
   * At least a TimeFormat or DateFormat should be defined, otherwise
   * an error occurs.
   *
   * @see https://momentjs.com/docs/#/displaying/format/
   */
  timeFormat: TimeFormat;

  /**
   * Optionally the locale moment js should use.
   */
  locale?: string;

  /**
   * Optional Callback which returns whether a date is selectable.
   * Is ran for every date which is displayed. By default, every
   * date can be selected.
   */
  isDateAllowed?: DateTimeInputIsDateAllowed;

  /**
   * When true, input time values will be interpreted as UTC (Zulu time)
   * by Moment.js. Otherwise they will default to the user's local
   * timezone.
   *
   * Defaults to true.
   */
  utc?: boolean;

  /**
   * Optionally customized text within the component.
   * This text should already be translated.
   */
  text?: Text;
};

export function DateTimeModal(props: Props) {
  const {
    onClose,
    onSave,
    label,
    dateFormat,
    timeFormat,
    locale,
    utc = true,
    text = {
      save: t({
        key: 'DateTimeModal.SELECT',
        fallback: 'Select'
      })
    }
  } = props;
  const isDateAllowed = get(props, 'isDateAllowed', constant(true));

  const [value, setValue] = useState(
    props.defaultValue ? moment(props.defaultValue) : ''
  );

  return (
    <OpenCloseModal
      onClose={onClose}
      onSave={() => onSave(value)}
      label={label}
      text={text}
      size="sm"
      className="date-time-modal"
      stickyFooter={false}
    >
      <Datetime
        input={false}
        open={true}
        closeOnSelect={false}
        onChange={setValue}
        value={value}
        dateFormat={dateFormat}
        timeFormat={timeFormat}
        locale={locale}
        utc={utc}
        isValidDate={(date: Moment, current?: Moment) =>
          isDateAllowed(date, current)
        }
      />
    </OpenCloseModal>
  );
}
