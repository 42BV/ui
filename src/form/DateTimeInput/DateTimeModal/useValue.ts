import { useState } from 'react';
import { Moment } from 'moment';

export function useValue(defaultValue: Moment | string) {
  return useState<Moment | string>(defaultValue);
}
