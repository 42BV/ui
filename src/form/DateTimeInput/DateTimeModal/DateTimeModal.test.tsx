import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { DateTimeModal } from './DateTimeModal';

import moment from 'moment';

describe('Component: DateTimeModal', () => {
  let console;

  afterEach(() => {
    global.console = console;
  });

  function setup({ hasValue }: { hasValue?: boolean }) {
    const onCloseSpy = jest.fn();
    const onSaveSpy = jest.fn();
    const isDateAllowedSpy = jest.fn().mockReturnValue(true);

    console = global.console;
    // @ts-expect-error We only use error, not all the other properties
    global.console = { error: jest.fn() };

    const { container } = render(
      <DateTimeModal
        onClose={onCloseSpy}
        onSave={onSaveSpy}
        dateFormat="YYYY-MM-DD"
        timeFormat="HH:II:SS"
        isDateAllowed={isDateAllowedSpy}
        defaultValue={hasValue ? new Date(2020, 1, 1, 0, 0, 0) : undefined}
      />
    );
    return { container, onCloseSpy, onSaveSpy, isDateAllowedSpy };
  }

  describe('ui', () => {
    test('default', () => {
      setup({ hasValue: true });
      expect(document.body.lastChild).toMatchSnapshot();
    });

    test('with value', () => {
      setup({ hasValue: true });
      expect(screen.getAllByText('1')[0]).toHaveClass('rdtActive');
    });

    test('without value', () => {
      setup({});
      const today = new Date();
      expect(screen.getAllByText(today.getDate()).map((e) => e.className)).toContain('rdtDay rdtToday');
    });
  });

  describe('events', () => {
    it('should update internal value when a date is selected', () => {
      const setValueSpy = jest.fn();
      jest.spyOn(React, 'useState').mockReturnValue([ '', setValueSpy ]);
      setup({});

      const value = moment(new Date).startOf('month');

      fireEvent.click(screen.getAllByText('1')[0]);

      expect(setValueSpy).toBeCalled();
      expect(value.isSame(setValueSpy.mock.calls.pop()[0])).toBe(true);
    });

    it('should not update external value when a date is selected', () => {
      const { onCloseSpy, onSaveSpy } = setup({});

      fireEvent.click(screen.getAllByText('1')[0]);

      expect(onCloseSpy).toBeCalledTimes(0);
      expect(onSaveSpy).toBeCalledTimes(0);
    });

    it('should close modal without value when modal is closed', () => {
      const { onCloseSpy, onSaveSpy } = setup({});

      fireEvent.click(screen.getAllByText('1')[0]);
      fireEvent.click(screen.getByText('cancel'));

      expect(onCloseSpy).toBeCalledTimes(1);
      expect(onSaveSpy).toBeCalledTimes(0);
    });

    it('should close modal with new value when clicking select button', () => {
      const { onCloseSpy, onSaveSpy } = setup({});

      const value = moment(new Date).startOf('month');

      fireEvent.click(screen.getAllByText('1')[0]);
      fireEvent.click(screen.getByText('save'));

      expect(onSaveSpy).toBeCalledTimes(1);
      expect(value.isSame(onSaveSpy.mock.calls[0][0])).toBe(true);

      expect(onCloseSpy).toBeCalledTimes(0);
    });

    test('is date allowed', () => {
      const isDateAllowedSpy = jest.fn().mockImplementation((date) => date.isBefore(new Date()));

      render(
        <DateTimeModal
          onClose={jest.fn()}
          onSave={jest.fn()}
          dateFormat="YYYY-MM-DD"
          timeFormat="HH:II:SS"
          isDateAllowed={isDateAllowedSpy}
        />
      );

      expect(isDateAllowedSpy).toBeCalledTimes(42);
    });
  });
});
