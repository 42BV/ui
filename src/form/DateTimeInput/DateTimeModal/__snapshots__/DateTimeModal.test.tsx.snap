// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DateTimeModal ui default 1`] = `
<div
  style="position: relative; z-index: 1050; display: block;"
  tabindex="-1"
>
  <div
    class=""
  >
    <div
      class="modal fade"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog date-time-modal modal-sm"
        role="document"
      >
        <div
          class="modal-content"
        >
          <form>
            <div
              class="modal-body"
            >
              <div
                class="rdt rdtStatic rdtOpen"
              >
                <div
                  class="rdtPicker"
                >
                  <div
                    class="rdtDays"
                  >
                    <table>
                      <thead>
                        <tr>
                          <th
                            class="rdtPrev"
                          >
                            <span>
                              ‹
                            </span>
                          </th>
                          <th
                            class="rdtSwitch"
                            colspan="5"
                            data-value="1"
                          >
                            February 2020
                          </th>
                          <th
                            class="rdtNext"
                          >
                            <span>
                              ›
                            </span>
                          </th>
                        </tr>
                        <tr>
                          <th
                            class="dow"
                          >
                            Su
                          </th>
                          <th
                            class="dow"
                          >
                            Mo
                          </th>
                          <th
                            class="dow"
                          >
                            Tu
                          </th>
                          <th
                            class="dow"
                          >
                            We
                          </th>
                          <th
                            class="dow"
                          >
                            Th
                          </th>
                          <th
                            class="dow"
                          >
                            Fr
                          </th>
                          <th
                            class="dow"
                          >
                            Sa
                          </th>
                        </tr>
                      </thead>
                      <tbody>
                        <tr>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="26"
                            data-year="2020"
                          >
                            26
                          </td>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="27"
                            data-year="2020"
                          >
                            27
                          </td>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="28"
                            data-year="2020"
                          >
                            28
                          </td>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="29"
                            data-year="2020"
                          >
                            29
                          </td>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="30"
                            data-year="2020"
                          >
                            30
                          </td>
                          <td
                            class="rdtDay rdtOld"
                            data-month="0"
                            data-value="31"
                            data-year="2020"
                          >
                            31
                          </td>
                          <td
                            class="rdtDay rdtActive"
                            data-month="1"
                            data-value="1"
                            data-year="2020"
                          >
                            1
                          </td>
                        </tr>
                        <tr>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="2"
                            data-year="2020"
                          >
                            2
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="3"
                            data-year="2020"
                          >
                            3
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="4"
                            data-year="2020"
                          >
                            4
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="5"
                            data-year="2020"
                          >
                            5
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="6"
                            data-year="2020"
                          >
                            6
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="7"
                            data-year="2020"
                          >
                            7
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="8"
                            data-year="2020"
                          >
                            8
                          </td>
                        </tr>
                        <tr>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="9"
                            data-year="2020"
                          >
                            9
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="10"
                            data-year="2020"
                          >
                            10
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="11"
                            data-year="2020"
                          >
                            11
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="12"
                            data-year="2020"
                          >
                            12
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="13"
                            data-year="2020"
                          >
                            13
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="14"
                            data-year="2020"
                          >
                            14
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="15"
                            data-year="2020"
                          >
                            15
                          </td>
                        </tr>
                        <tr>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="16"
                            data-year="2020"
                          >
                            16
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="17"
                            data-year="2020"
                          >
                            17
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="18"
                            data-year="2020"
                          >
                            18
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="19"
                            data-year="2020"
                          >
                            19
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="20"
                            data-year="2020"
                          >
                            20
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="21"
                            data-year="2020"
                          >
                            21
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="22"
                            data-year="2020"
                          >
                            22
                          </td>
                        </tr>
                        <tr>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="23"
                            data-year="2020"
                          >
                            23
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="24"
                            data-year="2020"
                          >
                            24
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="25"
                            data-year="2020"
                          >
                            25
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="26"
                            data-year="2020"
                          >
                            26
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="27"
                            data-year="2020"
                          >
                            27
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="28"
                            data-year="2020"
                          >
                            28
                          </td>
                          <td
                            class="rdtDay"
                            data-month="1"
                            data-value="29"
                            data-year="2020"
                          >
                            29
                          </td>
                        </tr>
                        <tr>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="1"
                            data-year="2020"
                          >
                            1
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="2"
                            data-year="2020"
                          >
                            2
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="3"
                            data-year="2020"
                          >
                            3
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="4"
                            data-year="2020"
                          >
                            4
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="5"
                            data-year="2020"
                          >
                            5
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="6"
                            data-year="2020"
                          >
                            6
                          </td>
                          <td
                            class="rdtDay rdtNew"
                            data-month="2"
                            data-value="7"
                            data-year="2020"
                          >
                            7
                          </td>
                        </tr>
                      </tbody>
                      <tfoot>
                        <tr>
                          <td
                            class="rdtTimeToggle"
                            colspan="7"
                          >
                            00:II:00
                          </td>
                        </tr>
                      </tfoot>
                    </table>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <span
                class="button d-inline-block ms-1 secondary"
              >
                <button
                  class="btn btn-secondary btn-md"
                  type="button"
                >
                  <div
                    class="d-flex justify-content-center align-items-center"
                  >
                    <i
                      class="icon button-icon me-2 material-icons"
                    >
                      cancel
                    </i>
                    Cancel
                  </div>
                </button>
              </span>
              <span
                class="button d-inline-block ms-1 primary"
              >
                <button
                  class="btn btn-primary btn-md"
                  type="submit"
                >
                  <div
                    class="d-flex justify-content-center align-items-center"
                  >
                    <i
                      class="icon button-icon me-2 material-icons"
                    >
                      save
                    </i>
                    Select
                  </div>
                </button>
              </span>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="modal-backdrop fade"
    />
  </div>
</div>
`;
